usersimcrs/simulator/neural/tus/tus_feature_handler.py

```diff
@@ -0,0 +1,516 @@
+        information need and mentioned during the conversation.
+
+        Args:
+            agent_dialogue_acts: Agent dialogue acts.
+            previous_state: Previous state.
+            state: Current state.
+            information_need: Information need.
+            user_action_vectors: User action feature vectors per slot. Defaults
+              to an empty dictionary.
+
+        Returns:
+            Feature vector for the turn.
+        """
+
+        # Update the action slots with constraints, requested, and mentioned
+        # slots
+        self.update_action_slots(information_need.constraints.keys())
+        self.update_action_slots(information_need.requested_slots.keys())
+        self.update_action_slots(
+            [
+                annotation.slot
+                for dialogue_act in agent_dialogue_acts
+                for annotation in dialogue_act.annotations
+            ]
+        )
+        return [
+            self.get_slot_feature_vector(
+                slot,
+                previous_state,
+                state,
+                information_need,
+                agent_dialogue_acts,
+                user_action_vectors.get(slot, None),
+            )
+            for slot in self.action_slots
+        ]
+
+    def _get_special_token_feature_vector(self, token: str) -> FeatureVector:
+        """Builds the feature vector for a special token.
+
+        Args:
+            token: Special token, either "[CLS]" or "[SEP]".
+
+        Raises:
+            ValueError: If the token is not supported.
+
+        Returns:
+            Feature vector for the special token.
+        """
+        if token not in ["[CLS]", "[SEP]"]:
+            raise ValueError(
+                f"Unsupported special token: {token}. Supported tokens: [CLS], "
+                "[SEP]"
+            )
+
+        vector = [1, 0] if token == "[CLS]" else [0, 1]
+        vector += [0] * 12  # Dimension of basic information feature
+        vector += [0] * len(self.slot_index)  # Dimension of slot index feature
+        vector += (
+            [0] * 3 * len(self._agent_actions)
+        )  # Dimension of agent action feature
+        vector += [0] * 6  # Dimension of user action feature
+        return vector
+
+    def build_input_vector(
+        self,
+        agent_dialogue_acts: List[DialogueAct],
+        previous_state: DialogueState,
+        state: DialogueState,
+        information_need: InformationNeed,
+        user_action_vectors: Dict[str, torch.Tensor] = {},
+    ) -> Tuple[List[FeatureVector], FeatureMask]:
+        """Builds the input vector $V_{input}$ for a turn.
+
+        It concatenates the feature vectors for the last n turns separated by
+        a special token. Note that is inferred from the list of feature vectors
+        provided.
+        The input vector is structured as follows:
+        [CLS] $V^t$ [SEP] $V^{t-1}$ [SEP] ... [SEP] $V^{t-n}$ [SEP] [PAD]
```

Comment:
The meaning of [PAD] is not explained anywhere. From here it appears as a special token, but it's actually padding, so perhaps indicate it differently (e.g., {padding})?

Reply:
Done.